src/cpp/server_core/ServerDatabase.cpp

```diff
@@ -0,0 +1,204 @@
+/*
+ * ServerDatabase.cpp
+ *
+ * Copyright (C) 2020 by RStudio, PBC
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+#include <server_core/ServerDatabase.hpp>
+
+#include <boost/algorithm/string.hpp>
+#include <boost/thread.hpp>
+
+#include <core/Log.hpp>
+#include <core/Settings.hpp>
+#include <core/system/Environment.hpp>
+#include <core/system/System.hpp>
+#include <shared_core/Error.hpp>
+
+namespace rstudio {
+namespace server_core {
+namespace database {
+
+using namespace core;
+using namespace core::database;
+
+namespace {
+
+// settings constants
+constexpr const char* kDatabaseProvider = "provider";
+constexpr const char* kDatabaseProviderSqlite = "sqlite";
+constexpr const char* kDatabaseProviderPostgresql = "postgresql";
+constexpr const char* kSqliteDatabaseDirectory = "directory";
+constexpr const char* kDefaultSqliteDatabaseDirectory = "/var/run/rstudio-server";
+constexpr const char* kDatabaseHost = "host";
+constexpr const char* kDefaultDatabaseHost = "localhost";
+constexpr const char* kDatabasePort = "port";
+constexpr const char* kDefaultPostgresqlDatabasePort = "5432";
+constexpr const char* kDatabaseUser = "user";
+constexpr const char* kDefaultPostgresqlDatabaseUser = "postgres";
+constexpr const char* kDatabasePassword = "password";
+constexpr const char* kPostgresqlDatabaseConnectionTimeoutSeconds = "connnection-timeout-seconds";
+constexpr const int   kDefaultPostgresqlDatabaseConnectionTimeoutSeconds = 10;
+
+// environment variables
+constexpr const char* kServerTmpDirEnvVar = "RS_SERVER_TMP_DIR";
+constexpr const char* kDatabaseMigrationsPathEnvVar = "RS_DB_MIGRATIONS_PATH";
+
+//misc constants
+constexpr const size_t kDefaultConnectionPoolSize = 4;
+
+boost::shared_ptr<ConnectionPool> s_connectionPool;
+
+Error readOptions(ConnectionOptions* pOptions)
+{
+   FilePath optionsFile("/etc/rstudio/database.conf");
+   if (optionsFile.exists())
+   {
+      // the database configuration file can potentially contain sensitive information
+      // log a warning if permissions are too lax
+      FileMode fileMode;
+      Error error = optionsFile.getFileMode(fileMode);
+      if (error)
+      {
+         LOG_ERROR_MESSAGE("Could not determine file permissions for database configuration file: " +
```

Comment:
This seems useful enough to factor into a helper function somewhere (IIRC we need to ensure 0600 permissions on a variety of files like this)

Reply:
I think this really depends on what you want to do with the file / what you want to log, whether or not you want to exit to make this really generic. The real meat of it is in `getFileMode` at any rate. Will leave this unresolved to solicit feedback from others. If there's more desire to make a utility function for this, I can attempt to do so!